### ISSUE TITLE
cargo check works but rust analyzer fails

### ISSUE THREAD

**Issue by zhiqiangxu** [rust-lang/rust-analyzer#15952]
I'm using visual studio code, which uses `rust analyzer` for compiling rust code. 

I'm experiencing the problem that `cargo check` succeeds but `rust analyzer` complains with errors.

The project I'm compiling depends on llvm so the `PATH` variable needs to be properly set, but I've already set it with `rust-analyzer.server.extraEnv` in settings.json.

What can be the possible reasons?

**Comment by alibektas** [comment]
Can you share what the error is so that we can help you further?

**Comment by zhiqiangxu** [comment]
> Can you share what the error is so that we can help you further?

[This](https://github.com/subspace/subspace) is the project I'm building, basically you need to install llvm and update the `PATH` variable following instructions [here](https://github.com/subspace/subspace/blob/main/docs/development.md#macos).

This is the error reported in vs code:

<img width="1657" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/1265027/13d9a42e-8b27-4a61-805d-caba8f97a9b5">

But `cargo check` reports ok

<img width="940" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/1265027/b6548eda-ab47-4dd0-8659-879643bfdd22">

**Comment by alibektas** [comment]
Interesting. I don't see the errors you see when I run it on my computer. Do you mind doing a `cargo clean`, I know the project is a heavy one but nothing else comes to mind.

**Comment by lnicola** [comment]
If it doesn't, you might have different environment variables in the terminals and Code. Try running Code from the terminal to see if things improve.

**Comment by zhiqiangxu** [comment]
> If it doesn't, you might have different environment variables in the terminals and Code. Try running Code from the terminal to see if things improve.

I noticed this because vs code will take quite some time to enable code navigation even though `cargo build` is already done. Maybe code navigation has nothing to do with `cargo build`?

**Comment by lnicola** [comment]
It doesn't, unless it says in the status bar it's building stuff.

Anyway, you might want to disable `rust-analyzer.check.allTargets`.

**Comment by zhiqiangxu** [comment]
The error is gone this time, after doing `cargo clean`.

But code navigation is still disabled.
<img width="1642" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/1265027/f7534fb3-1538-48da-adb7-c5cd81c4c87d">

**Comment by zhiqiangxu** [comment]
> It doesn't, unless it says in the status bar it's building stuff.
> 
> Anyway, you might want to disable `rust-analyzer.check.allTargets` to make it match `cargo check` in the terminal (but don't forget to actually run it from the terminal as said above).

Why is it necessary to disable `rust-analyzer.check.allTargets` ?

I think `cargo check` also checks all targets:

<img width="1015" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/1265027/0c16a05e-6bf7-4095-a542-3099c4b9f278">

**Comment by lnicola** [comment]
It makes `cargo` check the tests and examples too. It's the only difference between what we're running and `cargo check` in the terminal.

**Comment by zhiqiangxu** [comment]
> It's badly-named, it makes `cargo` check the tests and examples too. It's the only difference between what we're running and `cargo check` in the terminal.

Oh I see. Does code navigation work at your side?

**Comment by zhiqiangxu** [comment]
If rust-analyzer is already using the same build cache as `cargo build`, why is `cargo clean` needed in the first place, say when `cargo build` is ok?

**Comment by zhiqiangxu** [comment]
I tried `rust-analyzer.check.allTargets` and it's much faster now, but if I navigate to another crate, it will start showing loading again. Is it true that it has to rebuild the index after navigating to another file?

**Comment by lnicola** [comment]
> Is it true that it has to rebuild the index after navigating to another file?

For me, it does, because I have `rust-analyzer.cachePriming.enable` disabled. But you have so much code you might be blowing the caches. You could try increasing `rust-analyzer.lru.capacity` to something much larger.

**Comment by zhiqiangxu** [comment]
Things are much clearer now, thanks a lot @lnicola !

I think if rust-analyzer can show what it's doing in the background when it shows loading, the user experience will be much greater to users!

**Comment by lnicola** [comment]
It can't show what "Loading" in that hover pop-up means, but it's mostly doing type inference and trait solving.

You can add something like

```json
    "rust-analyzer.server.extraEnv": {
        "RA_PROFILE": "*>5000",
    }
```

to the settings, but it's not going to show anything too interesting. E.g.:

```
16631ms - handle_hover
    16629ms - classify_name_ref @ hash
        16629ms - Semantics::analyze_impl
            16629ms - infer:wait @ initialize_archiver
                0   - SourceBinder::to_module_def (1 calls)
                0   - crate_def_map:wait (1 calls)
        0   - SourceBinder::to_module_def (1 calls)
        0   - body_with_source_map_query (1 calls)
        0   - crate_def_map:wait (28 calls)
        0   - descend_into_macros (1 calls)
        0   - parse_query (1 calls)
```

You code, like every other Rust blockchain codebase, is very trait-heavy, which makes performance pretty bad. I mean parts like:

```rust
fn initialize_archiver<Block, Client, AS>(
    segment_headers_store: &SegmentHeadersStore<AS>,
    subspace_link: &SubspaceLink<Block>,
    client: &Client,
) -> sp_blockchain::Result<InitializedArchiver<Block>>
where
    Block: BlockT,
    Client: ProvideRuntimeApi<Block> + BlockBackend<Block> + HeaderBackend<Block> + AuxStore,
    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
    AS: AuxStore,
{ ... }

pub fn create_subspace_archiver<Block, Backend, Client, AS, SO>(
    segment_headers_store: SegmentHeadersStore<AS>,
    subspace_link: &SubspaceLink<Block>,
    client: Arc<Client>,
    sync_oracle: SubspaceSyncOracle<SO>,
    telemetry: Option<TelemetryHandle>,
) -> sp_blockchain::Result<impl Future<Output = sp_blockchain::Result<()>> + Send + 'static>
where
    Block: BlockT,
    Backend: BackendT<Block>,
    Client: ProvideRuntimeApi<Block>
        + BlockBackend<Block>
        + HeaderBackend<Block>
        + LockImportRun<Block, Backend>
        + Finalizer<Block, Backend>
        + AuxStore
        + Send
        + Sync
        + 'static,
    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
    AS: AuxStore + Send + Sync + 'static,
    SO: SyncOracle + Send + Sync + 'static,
{ ... }
```

**Comment by zhiqiangxu** [comment]
Is it possible to show the progress then? It's better than simply loading.

**Comment by lnicola** [comment]
No, but with cache priming enabled and a high enough LRU cache size, you shouldn't see them too often.

**Comment by zhiqiangxu** [comment]
<img width="1668" alt="image" src="https://github.com/rust-lang/rust-analyzer/assets/1265027/8817c6e8-9209-46a5-9502-f3e3a28b2fb0">
I `cmd+click` on `start_subspace`, and it seems rust-analyzer takes forever to find its caller in the red box.

**Comment by ChayimFriedman2** [comment]
It seems there isn't really anything actionable here.
